controller/instance_manager_controller.go

```diff
@@ -2534,3 +2546,55 @@ func (m *InstanceManagerMonitor) categorizeProcesses(instanceMap instanceProcess
 func (imc *InstanceManagerController) isResponsibleFor(im *longhorn.InstanceManager) bool {
 	return isControllerResponsibleFor(imc.controllerID, imc.ds, im.Name, im.Spec.NodeID, im.Status.OwnerID)
 }
+
+func isEngineInTransitionState(engine *longhorn.Engine) bool {
+	if !engine.DeletionTimestamp.IsZero() {
+		return true
+	}
+
+	spec := &engine.Spec
+	status := &engine.Status
+	if status.OwnerID == "" || status.CurrentState == "" {
+		// status has not been reconciled yet
+		return true
+	}
+
+	if status.CurrentState != spec.DesireState {
+		// not in desired state
+		return true
+	}
+
+	isAttached := spec.NodeID != ""
+	isScheduledToOwner := spec.NodeID == status.OwnerID
+	if isAttached && !isScheduledToOwner {
+		// attached but not on desired node
```

Comment:
It's better to add a comment explaining in which scenario the engine will fall into this status